contracts/contracts/interfaces/staking-manager.md

```diff
@@ -0,0 +1,116 @@
+- (?) Message is from P-Chain
+
+TODO: we might not want last 3 checks (with ?) to minimize reverts. The contract should be able to handle these cases without reverting. This is because the timing of the messages is not guaranteed and the contract should be able to handle out-of-order messages.
+
+If `RegisterSubnetValidatorTx` is not accepted, another warp message `InvalidValidatorRegisterMessage` will be relayed to the contract.
+
+## Remove Validator
+
+### removeValidator
+
+The manager contract can remove a validator gracefully by sending a warp message to P-Chain. `removeValidator` function should be called with the following inputs:
+
+- Subnet ID (bytes32): The ID of the subnet where the validator is being removed
+- Node ID (bytes32): The ID of the node that is being removed as a validator
+
+The contract should verify these:
+
+- The node is already registered as a validator
+- The address that registered the node is the same with the address that is removing the validator
+- (Optional) Validator being removed after minimum staking period
+
+After verification contract should perform the following:
+
+- Send a warp message to P-Chain for `SetSubnetValidatorWeightTx{messageID: relatedRegisterValidatorMessageID, nonce: MaxUint64, Weight: 0, }` transaction
+- Stop accruing rewards for the validator
+
+### receiveRemoveValidator
+
+Once the `SetSubnetValidatorWeightTx` is accepted, a warp message `InvalidValidatorRegisterMessage` will be relayed to the contract with `messageID` of the request. The message represents a messageID will not possibly become a valid validator. This essentialy means the message was not valid in the first place, or the validator has removed from the set. The InvalidValidatorRegisterMessage can either be a result of `removeValidator` from the contract or it can be originated directly from P-chain via `ExitValidatorSetTx`. In either case this ultimately tells the contract that a validator was evicted. See related ACP-77 section [here](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#exitvalidatorsettx).
+
+The contract should verify these in `receiveRemoveValidator`:
+
+- Verified warp message
+- The typecheck on the warp message (== InvalidValidatorRegisterMessage)
+- messageID is not empty
+
+At this point the contract should unlock all funds that were locked for the validator and issue locked rewards. If the validator was not removed by the contract itself, the contract should either give partial rewards or no rewards at all. This is because we don't know when the validator was removed.
+
+Rewards should not be unlocked or sent to the staker at this point. This is because we need to first perform a uptime check to determine the rewards.
+
+TODO: should we give a partial reward in case the validator got removed from P-chain (balance drained)? Possibly by timestamping the `InvalidValidatorRegisterMessage`.
+
+TODO: should we add a cooldown period for validator removal/readd?
```

Comment:
I was thinking if it would be DoSable if someone removes and readds itself too quickly. but I think that should be good. We should however still need to add a cooldown for partial redeeming.